extendeddatacrossword.go

```diff
@@ -94,87 +109,128 @@ func (eds *ExtendedDataSquare) solveCrossword(rowRoots [][]byte, columnRoots [][
+						switch mode {
+						case row:
+							r = i
+							c = j
 							vectorData = eds.Row(uint(i))
-						} else if mode == column {
-							rowIdx = j
-							colIdx = i
+						case column:
+							r = j
+							c = i
 							vectorData = eds.Column(uint(i))
+						default:
+							panic(fmt.Sprintf("invalid mode %d", mode))
 						}
-						if bitMask.Get(rowIdx, colIdx) {
+						if bitMask.Get(r, c) {
 							// As guaranteed by the bitMask, vectorData can't be nil here:
 							shares[j] = vectorData[j]
 						}
 					}
 
 					// Attempt rebuild
 					rebuiltShares, err := Decode(shares, eds.codec)
-					if err != nil { // repair unsuccessful
+					if err != nil {
+						// repair unsuccessful
 						solved = false
-					} else { // repair successful
-						progressMade = true
-						// Insert rebuilt shares into square
-						for p, s := range rebuiltShares {
-							if mode == row {
-								eds.setCell(uint(i), uint(p), s)
-							} else if mode == column {
-								eds.setCell(uint(p), uint(i), s)
-							}
-						}
-						if isExtendedPartIncomplete {
-							err := eds.rebuildExtendedPart(mode, uint(i))
-							if err != nil {
-								return err
-							}
-						}
+						continue
+					}
 
-						// Check that rebuilt vector matches given merkle root
-						err := eds.verifyRoots(rowRoots, columnRoots, mode, uint(i))
+					progressMade = true
+
+					if isExtendedPartIncomplete {
+						// If needed, rebuild the parity shares too.
+						rebuiltExtendedShares, err := Encode(rebuiltShares[0:eds.originalDataWidth], eds.codec)
 						if err != nil {
 							return err
 						}
+						startIndex := len(rebuiltExtendedShares) - int(eds.originalDataWidth)
+						rebuiltShares = append(
+							rebuiltShares[0:eds.originalDataWidth],
+							rebuiltExtendedShares[startIndex:]...,
+						)
+					} else {
+						// Otherwise copy them from the EDS.
+						startIndex := len(shares) - int(eds.originalDataWidth)
+						rebuiltShares = append(
+							rebuiltShares[0:eds.originalDataWidth],
+							shares[startIndex:]...,
+						)
+					}
 
-						// Check that newly completed orthogonal vectors match their new merkle roots
-						for j := 0; j < int(eds.width); j++ {
-							if mode == row && !bitMask.Get(i, j) {
-								if bitMask.ColumnIsOne(j) && !bytes.Equal(eds.ColRoot(uint(j)), columnRoots[j]) {
-									return &ErrByzantineColumn{uint(j)}
+					// Check that rebuilt shares matches appropriate root
+					err = eds.verifyAgainstRoots(rowRoots, columnRoots, mode, uint(i), bitMask, rebuiltShares)
+					if err != nil {
+						return err
+					}
+
+					// Check that newly completed orthogonal vectors match their new merkle roots
+					for j := 0; j < int(eds.width); j++ {
+						switch mode {
+						case row:
+							if !bitMask.Get(i, j) &&
+								bitMask.ColumnIsOne(j) {
+								err := eds.verifyAgainstRoots(rowRoots, columnRoots, column, uint(j), bitMask, rebuiltShares)
+								if err != nil {
+									return err
 								}
```

Comment:
We should aim for full coverage on the non-happy path branches too. (also orthogonal to this PR)